pkg/virt-controller/watch/vmi.go

```diff
@@ -1284,40 +1296,114 @@ func (c *VMIController) handleHotplugVolumes(hotplugVolumes []*virtv1.Volume, ho
+				logger.V(3).Infof("Volume %s/%s is in WaitForFistConsumer, triggering population", vmi.Namespace, volume.Name)
+				syncError := c.triggerHotplugPopulation(volume, vmi, virtLauncherPod)
+				if syncError != nil {
+					return syncError
+				}
+				continue
+			}
+			if !ready {
+				// Volume not ready, skip until it is.
+				logger.V(3).Infof("Skipping hotplugged volume: %s, not ready", volume.Name)
+				continue
+			}
+			// Check if the VMI VolumeStatus contains this volume, if that is the case then something deleted the attachment pod
+			// and we need to stop the VMI as that is a critical error.
+			if c.volumeStatusContainsVolumeAndPod(vmi.Status.VolumeStatus, volume) {
+				logger.V(1).Infof("Detected attachment pod is missing for VMI %s/%s, the VMI will be deleted", vmi.Namespace, vmi.Name)
+				return &syncErrorImpl{fmt.Errorf("Missing pod for hotplugged volume %s", volume.Name), MissingAttachmentPodReason}
+			}
+			if err := c.createAttachmentPod(vmi, virtLauncherPod, volume); err != nil {
+				return err
+			}
 		}
+	}
+
+	return nil
+}
+
+func (c *VMIController) createAttachmentPod(vmi *virtv1.VirtualMachineInstance, virtLauncherPod *k8sv1.Pod, volume *virtv1.Volume) syncError {
+	attachmentPodTemplate, err := c.createAttachmentPodTemplate(vmi, virtLauncherPod, volume)
+	if attachmentPodTemplate == nil { // nil means the PVC is not populated yet.
+		return nil
+	}
+	vmiKey := controller.VirtualMachineKey(vmi)
+	c.podExpectations.ExpectCreations(vmiKey, 1)
+
+	pod, err := c.clientset.CoreV1().Pods(vmi.GetNamespace()).Create(context.TODO(), attachmentPodTemplate, v1.CreateOptions{})
+	if err != nil {
+		c.podExpectations.CreationObserved(vmiKey)
+		c.recorder.Eventf(vmi, k8sv1.EventTypeWarning, FailedCreatePodReason, "Error creating hotplug pod for volume %s: %v", volume.Name, err)
+		return &syncErrorImpl{fmt.Errorf("Error creating attachment pod %v", err), FailedCreatePodReason}
+	}
+	c.recorder.Eventf(vmi, k8sv1.EventTypeNormal, SuccessfulCreatePodReason, "Created attachment pod %s for volume %s", pod.Name, volume.Name)
+	return nil
+}
+
+func (c *VMIController) triggerHotplugPopulation(volume *virtv1.Volume, vmi *virtv1.VirtualMachineInstance, virtLauncherPod *k8sv1.Pod) syncError {
+	populateHotplugPodTemplate, err := c.createAttachmentPopulateTriggerPodTemplate(volume, virtLauncherPod, vmi)
+	if err != nil {
+		return &syncErrorImpl{fmt.Errorf("Error creating trigger pod template %v", err), FailedCreatePodReason}
+	}
+	if populateHotplugPodTemplate != nil { // nil means the PVC is not populated yet.
 		vmiKey := controller.VirtualMachineKey(vmi)
 		c.podExpectations.ExpectCreations(vmiKey, 1)
 
-		pod, err := c.clientset.CoreV1().Pods(vmi.GetNamespace()).Create(context.Background(), attachmentPodTemplate, v1.CreateOptions{})
+		_, err := c.clientset.CoreV1().Pods(vmi.GetNamespace()).Create(context.TODO(), populateHotplugPodTemplate, v1.CreateOptions{})
```

Comment:
That is a good question, I never actually noticed KubeVirt uses a lot of context.BackGround. I had assumed basically everyone used context.TODO, unless there was a specific reason not to. This is taken directly from context.go:
```go
// Background returns a non-nil, empty Context. It is never canceled, has no
// values, and has no deadline. It is typically used by the main function,
// initialization, and tests, and as the top-level Context for incoming
// requests.
func Background() Context {
	return background
}

// TODO returns a non-nil, empty Context. Code should use context.TODO when
// it's unclear which Context to use or it is not yet available (because the
// surrounding function has not yet been extended to accept a Context
// parameter).
func TODO() Context {
	return todo
}
```
I am not quite sure if either one makes much of a difference. I mostly use context.TODO because its less characters.

Reply:
I grepped the code base for both, and it does appear KubeVirt uses mostly context.Background, but there are a lot of context.TODOs in there.